### ISSUE TITLE
Design Review 2019-02-27 16:30 UTC (Africa/Europe/western Asia)

### ISSUE THREAD

**Issue by mrjoro** [ampproject/amphtml#20537]
Time: [2019-02-27 16:30 UTC](https://www.timeanddate.com/worldclock/meeting.html?year=2019&month=02&day=27&iv=0) ([add to Google Calendar](http://www.google.com/calendar/event?action=TEMPLATE&text=AMP%20Project%20Design%20Review&dates=20190227T163000Z/20190227T173000Z&details=https%3A%2F%2Fhangouts.google.com%2Fhangouts%2F_%2Fgoogle.com%2Famp-eng-desrev))
Location: [Video conference via Google Meet](https://bit.ly/amp-dr)

The AMP Project holds weekly engineering [design reviews](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md).  **We encourage everyone in the community to participate in these design reviews.**

If you are interested in bringing your design to design review, read the [design review documentation](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md) and add a link to your design doc by the Monday before your design review.

When attending a design review please read through the designs *before* the design review starts.  This allows us to spend more time on discussion of the design.

We rotate our design review between times that work better for different parts of the world as described in our [design review documentation](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md), but you are welcome to attend any design review.  If you cannot make any of the design reviews but have a design to discuss please let mrjoro@ know on [Slack](https://github.com/ampproject/amphtml/blob/master/CONTRIBUTING.md#discussion-channels) and we will find a time that works for you.

**Comment by caroqliu** [comment]
I'd like to discuss #20988, a new extension `<amp-autocomplete>`

**Comment by cvializ** [comment]
## amp-autocomplete

### Overview
- assist user to enter input more quickly, e.g. e-commerce, news
- use-cases: form data, search field
- amp-form integration
- main features
  - filtering mechanism
    - various predefined functions, or custom
    - a none option to accept all server data
  - data source
    - static inline
    - remote data
- other features
  - changing enter behavior
  - rich results templating
- fallbacks and placeholders
- various usage examples

### Questions
- (Alan Kent) use case for autocomplete, with product titles and images. can autocomplete work if product titles are listed with images? e.g. they don't know what the categories are while they're doing a search. showing related categories for the given query?
  - @caroqliu not a main goal for the feature but will keep in mind
  - @aghassemi Publisher can specify flexibly any data and template can branch on those data
- (Alan Kent) tokenized autocomplete, how will matches translate into submission?
  - @caroqliu tokenized autocomplete as a filter
  - @nainar Generic autocomplete should work when use types various results
  - (Alan Kent) Autocomplete can teach users about how to use the site and can improve first-time experience
- (Alan Kent) Live chat widget. Performance and caching: will this work when used with a single-line service worker. If a user makes the same query twice, will the query be able to load instantly. If user types "wordA wordB" then removed "wordB" will the request be made again or will there be an optimization? Can requests be preloaded?
  - @jridgewell Store all autocomplete results in a single LRU cache in memory (for all components on the page) and use that for caching previous requests. 
- @jpettitt grey-text vs highlighted-text, can we support both? 
  - @caroqliu UX for keypresses is simpler in the highlight text format.
  - @andrewwatterson mostly just a visual flourish, so we prioritize highlighted-text since it requires UX thinking.

**Comment by andrewwatterson** [comment]
So, I *think* Alan Kent was suggesting something somewhat different than what the design doc refers to as "tokenized autocomplete." I think he was suggesting a new option for `filter` that would break the user's input into tokens, perform `token-prefix` filtering on EACH of them, and then return the intersection of those results. For example, if I searched for "Kent, Alan," it would return "Alan Kent" because both tokens in my input match tokens in the result. In this way, it meets users who are expecting a different format (last name first) and teaches them the format the site naturally uses (first name first). I think that's a great suggestion - I'll let @nainar decide if it's MVP.

Also, we prioritized `highlight-text` because it materially changes the interaction (which is/sounds like a user benefit), not because it requires UX thinking (which sounds self-aggrandizing). :-P
